### PR TITLE
Add API to force an image refresh

### LINKED ISSUE #2358: Force an image update
Would be useful, I think, to have a way to trigger an image check & update. Maybe just for auto-update images. Something like "lxc image update <image-id|alias>". That would be quicker than waiting for the auto-update to kick in, and also quicker than deleting and re-creating the image.

### PR DESCRIPTION
**Pull request by albertodonato**
Add an API endpoint `image/<hash>/refresh` to require that an image is refreshed from its source.

It can be called with `lxc image refresh <name>`

This also makes the server save the image certificate when provided by the client and passes it to ImageDownload so that refresh works between LXDs too

Closes #2358 and #3298

### CODE REVIEW

**Comment by stgraber** [comment]
@albertodonato failed static analysis

File: client/lxd_images.go [hunk 1]
```diff
@@ -496,6 +496,17 @@ func (r *ProtocolLXD) DeleteImage(fingerprint string) (*Operation, error) {
 	return op, nil
 }
 
+// RefreshImage requests that LXD issues an image refresh
+func (r *ProtocolLXD) RefreshImage(fingerprint string) (*Operation, error) {
```
Comment (stgraber) [inline client/lxd_images.go]: This should have a r.HasExtension check for image_force_update

Comment (albertodonato) [inline client/lxd_images.go]: Done, I also renamed it to image_force_refesh

**Comment by stgraber** [comment]
```
# github.com/lxc/lxd/client
client/lxd_images.go:502: not enough arguments to return
```

**Comment by albertodonato** [comment]
Fixed

File: client/lxd_images.go [hunk 2]
```diff
@@ -496,6 +496,21 @@ func (r *ProtocolLXD) DeleteImage(fingerprint string) (*Operation, error) {
...
+	if !r.HasExtension("image_force_refresh") {
+		return nil, fmt.Errorf("The server is missing the required \"certificate_update\" API extension")
```
Comment (stgraber) [inline client/lxd_images.go]: Bad copy/paste in the error message.

**Comment by stgraber** [comment]
Should be ready to merge once that bad copy/paste is fixed.

File: client/lxd_images.go (hunk 2 above)
Comment (albertodonato) [inline client/lxd_images.go]: Oops, fixed

**Comment by kikoreis** [comment]
For the record, this is super useful for a laptop user like me who may not always be connected in the cheapest way (i.e. 4G WWAN etc) and would like to control image refreshing. Today I use

{{{lxc config set images.auto_update_cached false}}}

but there's no easy way to update the images then. So thanks!

**Comment by mpontillo** [comment]
Is it expected that this doesn't work with the standard Ubuntu image streams?

I was considering turning off auto-update of cached images and writing a cron job to refresh the images I want to use regularly, so it doesn't interrupt my workflow upon container creation (as the images can be a bit slow to download).

But when I try it with `ubuntu:<image>` I get:

```
$ lxc image refresh ubuntu:x
Error: The remote isn't a private LXD server
```

**Comment by stgraber** [comment]
@mpontillo the `lxc image refresh` must be targeted at a cached image in your local image store (either a fingerprint or a local alias).

**Comment by mpontillo** [comment]
@stgraber, thanks.

So that means if I want to write a script to keep an existing images up-to-date, I have to do the equivalent of `lxc image list --format yaml`, parse the output, look for the `update_source`, check if it matches the set of images I want to keep updated, (though realistically, I might want to just keep everything updated that has an `update_source`) and then issue the `lxc image refresh` command on that fingerprint.

I guess it would be nice if `lxc image refresh` could do that work itself; maybe `lxc image refresh --all` could update anything that has the ability to be updated.

**Comment by mpontillo** [comment]
Well, to be fair, I could do it the brute force way:

```
for fprint in $(lxc image list --format csv --columns f); do 
   lxc image refresh $fprint 2> /dev/null || true
done
```
... which would just ignore the errors for any images I have locally cached without an `update_source`.

**Comment by stgraber** [comment]
@mpontillo you could also just assign an alias to your local image and then use that to refresh it. It'd save you the fingerprint lookup.

**Comment by mpontillo** [comment]
That's a good idea; maybe I'll make a habit of running something like the following every time I initialize a new `lxd`:

```
lxc image copy ubuntu:b local: --copy-aliases
lxc image copy ubuntu:x local: --copy-aliases
lxc image copy ubuntu:t local: --copy-aliases
```

Then I can just refresh them by name as needed, and I don't have to worry about waiting for an image download the first time I launch a container.

For what it's worth, I love that running these commands after an image is already downloaded updates the aliases and doesn't re-download images that are already cached; nice! (I wonder if the aliases from the server are copied and appended to locally, or copied and overwritten locally, though.)

**Comment by albertodonato** [comment]
@mpontillo if all you want is to keep images up to date, you can set `auto_update: true` on the image, and LXD will periodically update.

E.g.:
```
$ lxc image show b
auto_update: true
properties:
  architecture: amd64
  description: ubuntu 18.04 LTS amd64 (daily) (20180809)
  label: daily
  os: ubuntu
  release: bionic
  serial: "20180809"
  version: "18.04"
public: false
```


The `refresh` command is meant for when you want to make sure you have the latest image right away.

**Pull request merged**